Unconfigured HDF plugin problems

- refs #1061 
- refs https://github.com/BCDA-APS/bluesky_training/issues/335
- refs https://github.com/BCDA-APS/bluesky_training/issues/311

Following https://github.com/BCDA-APS/bluesky_training/issues/311, encountered some problems when using an area detector with the HDF1 plugin as a Component.  Specifically, this code (in [`AD_EpicsFileNameMixin.stage()`](https://github.com/BCDA-APS/apstools/blob/fd1b8c44db65af46505e3c6da04c118e4aa3088b/apstools/devices/area_detector_support.py#L490-L495)) was still executed:

```py
        try:
            # assume template includes format for a file number
            self._fn = template % (read_path, filename, file_number)
        except (TypeError, ValueError):
            # in case template does not include file_number
            self._fn = template % (read_path, filename)
```

It failed when using a detector IOC that was just started with no additional configuration.  Both `file_template` and `file_path` contained default values.  Those values triggered exceptions which failed the run.

Need to examine if other tests are affected, like this one:

```py
FAILED apstools/devices/tests/test_area_detector_support.py::test_AD_EpicsFileNameMixin[hdf1-AD_HDF5] - ValueError: 'ad:JPEG1:FileName' is an empty string.
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -36,6 +36,7 @@ describe future plans.
    Fixes
    -----
 
+   * Fix area detector staging of unconfigured HDF5 plugin.
    * Fix problem with detector selection for lineup2() reported by USAXS.
    * Fix unit test involving package version string length.
 

--- a/apstools/devices/area_detector_factory.py
+++ b/apstools/devices/area_detector_factory.py
@@ -90,6 +90,7 @@ from .area_detector_support import AD_EpicsFileNameJPEGPlugin
 from .area_detector_support import AD_EpicsFileNameTIFFPlugin
 from .area_detector_support import BadPixelPlugin
 from .area_detector_support import HDF5FileWriterPlugin
+from .area_detector_support import SimDetectorCam_V34
 from .area_detector_support import SingleTrigger_V34
 
 DEFAULT_DETECTOR_BASES = (
@@ -101,7 +102,7 @@ PLUGIN_DEFAULTS = {  # some of the common plugins
     # gets image from the detector
     "cam": {
         "suffix": "cam1:",
-        "class": ophyd.areadetector.CamBase,
+        "class": SimDetectorCam_V34,
     },
     # for imaging
     "image": {

--- a/apstools/devices/area_detector_support.py
+++ b/apstools/devices/area_detector_support.py
@@ -38,11 +38,11 @@ from collections import OrderedDict
 
 import epics
 import numpy as np
+from ophyd import ADComponent
+from ophyd import CamBase
 from ophyd import EpicsSignal
 from ophyd import EpicsSignalRO
 from ophyd import EpicsSignalWithRBV
-from ophyd import ADComponent
-from ophyd import CamBase
 from ophyd import SimDetectorCam
 from ophyd import SingleTrigger
 from ophyd.areadetector.filestore_mixins import FileStoreBase
@@ -413,7 +413,7 @@ class AD_EpicsFileNameMixin(FileStorePluginBase):
     """
 
     def _remove_caller_stage_sigs(self):
-        """Caller is responsible for setting these stage_sigs."""
+        """Caller is responsible for setting these, pop from stage_sigs."""
         caller_sets_these = """
             array_counter
             auto_increment
@@ -427,8 +427,7 @@ class AD_EpicsFileNameMixin(FileStorePluginBase):
             num_capture
         """.split()
         for key in caller_sets_these:
-            if key in self.stage_sigs:
-                self.stage_sigs.pop(key)
+            self.stage_sigs.pop(key, None)
 
     def make_filename(self):
         """
@@ -464,6 +463,12 @@ class AD_EpicsFileNameMixin(FileStorePluginBase):
         Set EPICS items before device is staged, then copy EPICS
         naming template (and other items) to ophyd after staging.
         """
+        # Validate!
+        for attr in "file_name file_path file_template".split():
+            signal = getattr(self, attr)
+            if len(signal.get()) == 0:
+                raise ValueError(f"{signal._write_pv.pvname!r} is an empty string.")
+
         if "capture" in self.stage_sigs:
             self.stage_sigs.move_to_end("capture", last=True)
 

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -31,6 +31,13 @@ from .. import BadPixelPlugin
 from .. import ad_creator
 
 
+def caput_AD(signal, value):
+    """Low-level caput to AD Component (avoids failing Status objects)."""
+    import epics
+
+    epics.caput(signal._write_pv.pvname, value)
+
+
 def custom_ad_creator(plugin_name):
     """Create area detector with named file writer plugin."""
     plugin_class = dict(
@@ -61,7 +68,7 @@ def custom_ad_creator(plugin_name):
     return adsimdet
 
 
-def AD_unprime_plugin(plugin):
+def unprime_AD_plugin(plugin):
     """
     Make area detector plugin require priming again.
 
@@ -97,7 +104,7 @@ def AD_unprime_plugin(plugin):
                 auto_save=0,  # "No"
                 lazy_open="No",
                 create_directory=0,
-                file_write_mode=2, # "Single"
+                file_write_mode=2,  # "Single"
                 file_template="%s%s_%d.trouble",
                 file_path="/tmp/",
                 file_name="test",
@@ -137,11 +144,7 @@ def AD_unprime_plugin(plugin):
         ],
     ],
 )
-def test_replacement(plugin_name, spec, presets, prime, context, expected):
-    import epics
-
-    def caput(signal, value):
-        epics.caput(signal._write_pv.pvname, value)
+def test_AD_EpicsFileNameMixin(plugin_name, spec, presets, prime, context, expected):
 
     with context as exinfo:
         det = custom_ad_creator(plugin_name)
@@ -159,7 +162,7 @@ def test_replacement(plugin_name, spec, presets, prime, context, expected):
         plugin.stage_sigs.move_to_end("capture", last=True)
         assert list(plugin.stage_sigs)[-1] == "capture"
 
-        AD_unprime_plugin(plugin)
+        unprime_AD_plugin(plugin)
 
         if prime and not AD_plugin_primed(plugin):
             AD_prime_plugin2(plugin)
@@ -171,7 +174,7 @@ def test_replacement(plugin_name, spec, presets, prime, context, expected):
 
         for attr, value in settings.items():
             if attr in dir(plugin):
-                caput(getattr(plugin, attr), value)
+                caput_AD(getattr(plugin, attr), value)
 
         try:
             det.stage()
@@ -199,86 +202,6 @@ def test_replacement(plugin_name, spec, presets, prime, context, expected):
 
     if expected is not None:
         assert expected in str(exinfo)
-
-
-@pytest.mark.skipif(
-    in_gha_workflow(),
-    reason="Random failures in GiHub Actions workflows.",
-)
-@pytest.mark.parametrize(
-    "plugin_name, spec",
-    [
-        ["hdf1", "AD_HDF5"],
-        ["jpeg1", "AD_JPEG"],
-        ["tiff1", "AD_TIFF"],
-    ],
-)
-def test_AD_EpicsFileNameMixin(plugin_name, spec, adsimdet):
-    import epics
-
-    def caput(signal, value):
-        epics.caput(signal._write_pv.pvname, value)
-
-    # adsimdet = custom_ad_creator(plugin_name)
-    assert adsimdet is not None
-    assert plugin_name in dir(adsimdet)
-
-    # Preset some plugin values required for staging.
-    for key in "hdf jpeg tiff".split():
-        plugin = getattr(adsimdet, f"{key}1")
-        caput(plugin.create_directory, -5)
-        caput(plugin.file_name, f"test_{key}")
-        caput(plugin.file_path, f"/tmp/test_{key}/")
-        caput(plugin.file_template, f"%s%s_%3.3d.{key}/")
-
-    # basic plugin tests
-    plugin = getattr(adsimdet, plugin_name)
-    assert AD_EpicsFileNameMixin is not None
-    assert isinstance(plugin, AD_EpicsFileNameMixin)
-    assert "filestore_spec" in dir(plugin)
-    assert plugin.filestore_spec == spec
-    assert isinstance(plugin.stage_sigs, dict)
-    assert len(plugin.stage_sigs) >= 4
-    assert "capture" in plugin.stage_sigs
-
-    # configuration prescribed for the user
-    user_settings = dict(
-        array_counter=0,
-        auto_increment=1,  # Yes
-        auto_save=0,  # No
-        create_directory=-5,
-        file_template=f"%s%s_%2.2d.{plugin_name}",
-        # file_template before file_name, number, & path
-        file_name="flotsam",
-        file_number=1 + int(10 * random.random()),
-        file_path=f"{pathlib.Path(tempfile.mkdtemp())}/",  # ALWAYS ends with "/"
-        num_capture=1 + int(10 * random.random()),
-    )
-    if plugin_name == "hdf1":
-        user_settings["compression"] = "zlib"
-
-    # add the settings
-    for attr, value in user_settings.items():
-        caput(getattr(plugin, attr), value)
-    timed_pause(0.1)
-
-    adsimdet.stage()
-    if plugin_name == "hdf1":  # Why special case?
-        user_settings["file_number"] += 1  # the IOC will do the same
-    assert list(plugin.stage_sigs.keys())[-1] == "capture"
-    for k, v in user_settings.items():
-        assert getattr(plugin, k).get() == v, f"{plugin_name=} {k=}  {v=}"
-
-    assert plugin.get_frames_per_point() == user_settings["num_capture"]
-
-    filename, read_path, write_path = plugin.make_filename()
-    assert isinstance(filename, str)
-    assert isinstance(read_path, str)
-    assert isinstance(write_path, str)
-    assert filename == user_settings["file_name"]
-    assert read_path.startswith(datetime.datetime.now().strftime(READ_PATH_TEMPLATE))
-    assert write_path == user_settings["file_path"]
-    adsimdet.unstage()
 
 
 @pytest.mark.parametrize(
@@ -581,8 +504,6 @@ def test_plugin(Klass, setup, attrs, context, expected):
 )
 def test_HDF5plugin_i1062(setup, attrs, context, expected):
     """Issue #1062, Unconfigured HDF plugin."""
-    import epics
-
     with context as exinfo:
         det = ad_creator(IOC_AD, name="det", **setup)
 
@@ -595,7 +516,7 @@ def test_HDF5plugin_i1062(setup, attrs, context, expected):
             assert AD_plugin_primed(det.hdf1)
 
         for attr, value in attrs.items():
-            epics.caput(getattr(det.hdf1, attr)._write_pv.pvname, value)
+            caput_AD(getattr(det.hdf1, attr), value)
 
         RE = bluesky.RunEngine()
         (uid,) = RE(bp.count([det]))

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -89,7 +89,7 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec):
     user_settings = dict(
         array_counter=0,
         auto_increment=1,  # Yes
-        auto_save=1,  # Yes
+        auto_save=0,  # No
         create_directory=-5,
         file_name="flotsam",
         file_number=1 + int(10 * random.random()),
@@ -426,18 +426,6 @@ def test_HDF5plugin_i1062(setup, attrs, context, expected):
     """Issue #1062, Unconfigured HDF plugin."""
     import epics
 
-    # Caller is responsible for setting values of these Components:
-    # * array_counter
-    # * auto_increment
-    # * auto_save
-    # * compression (only HDF)
-    # * create_directory
-    # * file_name
-    # * file_number
-    # * file_path
-    # * file_template
-    # * num_capture
-
     with context as exinfo:
         det = ad_creator(IOC_AD, name="det", **setup)
 
@@ -445,28 +433,12 @@ def test_HDF5plugin_i1062(setup, attrs, context, expected):
         for attr in "file_name file_path file_template".split():
             assert attr not in det.hdf1.stage_sigs
 
-        # # un-prime the HDF5 plugin
-        # det.cam.data_type.put("Int8")
-        # AD_prime_plugin2(det.hdf1)
-        # det.cam.data_type.put("UInt8")
-        # time.sleep(0.1)
-        # assert not AD_plugin_primed(det.hdf1)
-
         if not AD_plugin_primed(det.hdf1):
             AD_prime_plugin2(det.hdf1)
             assert AD_plugin_primed(det.hdf1)
 
         for attr, value in attrs.items():
             epics.caput(getattr(det.hdf1, attr)._write_pv.pvname, value)
-
-        # # This happens during det.hdf1.stage()
-        # filename, read_path, write_path = det.hdf1.make_filename()
-        # # Test its results.
-        # formatter = datetime.datetime.now().strftime
-        # assert filename == det.hdf1.file_name.get()
-        # assert read_path == setup["plugins"][1]["hdf1"]["read_path_template"]
-        # assert read_path == formatter(det.hdf1.read_path_template)
-        # assert write_path == formatter(det.hdf1.file_path.get())
 
         RE = bluesky.RunEngine()
         (uid,) = RE(bp.count([det]))

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -61,6 +61,26 @@ def custom_ad_creator(plugin_name):
     return adsimdet
 
 
+def AD_unprime_plugin(plugin):
+    """
+    Make area detector plugin require priming again.
+
+    One way is to prime the plugin with a different data type.
+    """
+    cam = plugin.parent.cam
+    original_type = cam.data_type.get(use_monitor=False, as_string=True)
+    assert original_type != "Int8"
+
+    cam.data_type.put("Int8")  # change it
+    AD_prime_plugin2(plugin)
+    timed_pause(0.1)
+    assert AD_plugin_primed(plugin)
+
+    cam.data_type.put(original_type)  # restore it
+    timed_pause(0.1)
+    assert not AD_plugin_primed(plugin)
+
+
 @pytest.mark.parametrize(
     "plugin_name, spec",
     [
@@ -70,23 +90,54 @@ def custom_ad_creator(plugin_name):
     ],
 )
 @pytest.mark.parametrize(
-    "presets, context, expected",
+    "presets, prime, context, expected",
     [
         [
             dict(
-                auto_save="No",
+                auto_save=0,  # "No"
                 lazy_open="No",
-                create_directory=-5,
-                file_template="%s%s_%2.2d.ad_data",
+                create_directory=0,
+                file_write_mode=2, # "Single"
+                file_template="%s%s_%d.trouble",
                 file_path="/tmp/",
                 file_name="test",
+                file_number=1 + int(9 * random.random()),
             ),
+            False,
+            does_not_raise(),
+            None,
+        ],
+        [
+            dict(
+                auto_save=0,  # "No"
+                lazy_open="No",
+                create_directory=0,
+                file_template="",
+                file_path="",
+                file_name="",
+                file_number=1 + int(9 * random.random()),
+            ),
+            True,
+            pytest.raises(ValueError),
+            ":FileName' is an empty string.",
+        ],
+        [
+            dict(
+                auto_save=0,  # "No"
+                lazy_open="No",
+                create_directory=-5,
+                file_template="%s%s_%2.2d.ad_image",
+                file_path="/tmp/",
+                file_name="test",
+                file_number=1 + int(9 * random.random()),
+            ),
+            True,
             does_not_raise(),
             None,
         ],
     ],
 )
-def test_replacement(plugin_name, spec, presets, context, expected):
+def test_replacement(plugin_name, spec, presets, prime, context, expected):
     import epics
 
     def caput(signal, value):
@@ -104,10 +155,13 @@ def test_replacement(plugin_name, spec, presets, context, expected):
         assert len(plugin.stage_sigs) >= 4
         assert plugin.stage_sigs.get("file_write_mode") == "Stream"
         assert plugin.stage_sigs.get("capture") in (1, "Capture")
+
         plugin.stage_sigs.move_to_end("capture", last=True)
         assert list(plugin.stage_sigs)[-1] == "capture"
 
-        if not AD_plugin_primed(plugin):
+        AD_unprime_plugin(plugin)
+
+        if prime and not AD_plugin_primed(plugin):
             AD_prime_plugin2(plugin)
             assert AD_plugin_primed(plugin)
 
@@ -119,9 +173,29 @@ def test_replacement(plugin_name, spec, presets, context, expected):
             if attr in dir(plugin):
                 caput(getattr(plugin, attr), value)
 
-        RE = bluesky.RunEngine()
-        (uid,) = RE(bp.count([det]))
-        assert isinstance(uid, str)
+        try:
+            det.stage()
+
+            assert list(plugin.stage_sigs.keys())[-1] == "capture"
+            for k, v in settings.items():
+                assert getattr(plugin, k).get(use_monitor=False) == v, f"{plugin_name=} {k=}  {v=}"
+
+            # assert plugin.get_frames_per_point() == settings["num_capture"]
+
+            filename, read_path, write_path = plugin.make_filename()
+            assert isinstance(filename, str)
+            assert isinstance(read_path, str)
+            assert isinstance(write_path, str)
+            assert filename == settings["file_name"]
+            assert read_path.startswith(datetime.datetime.now().strftime(READ_PATH_TEMPLATE))
+            assert write_path == settings["file_path"]
+
+        finally:
+            det.unstage()
+
+        # RE = bluesky.RunEngine()
+        # (uid,) = RE(bp.count([det]))
+        # assert isinstance(uid, str)
 
     if expected is not None:
         assert expected in str(exinfo)

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -145,7 +145,7 @@ def unprime_AD_plugin(plugin):
     ],
 )
 def test_AD_EpicsFileNameMixin(plugin_name, spec, presets, prime, context, expected):
-
+    """Test the mixin with stage/unstage."""
     with context as exinfo:
         det = custom_ad_creator(plugin_name)
 
@@ -183,8 +183,6 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec, presets, prime, context, expec
             for k, v in settings.items():
                 assert getattr(plugin, k).get(use_monitor=False) == v, f"{plugin_name=} {k=}  {v=}"
 
-            # assert plugin.get_frames_per_point() == settings["num_capture"]
-
             filename, read_path, write_path = plugin.make_filename()
             assert isinstance(filename, str)
             assert isinstance(read_path, str)
@@ -195,10 +193,6 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec, presets, prime, context, expec
 
         finally:
             det.unstage()
-
-        # RE = bluesky.RunEngine()
-        # (uid,) = RE(bp.count([det]))
-        # assert isinstance(uid, str)
 
     if expected is not None:
         assert expected in str(exinfo)

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -93,8 +93,8 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec):
         auto_increment=1,  # Yes
         auto_save=0,  # No
         create_directory=-5,
-        # file_template before file_name & file_path
         file_template=f"%s%s_%2.2d.{plugin_name}",
+        # file_template before file_name, number, & path
         file_name="flotsam",
         file_number=1 + int(10 * random.random()),
         file_path=f"{pathlib.Path(tempfile.mkdtemp())}/",  # ALWAYS ends with "/"
@@ -107,7 +107,7 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec):
     for attr, value in user_settings.items():
         sig = getattr(plugin, attr)
         epics.caput(sig._write_pv.pvname, value)
-    timed_pause()
+    timed_pause(0.1)
 
     adsimdet.stage()
     if plugin_name == "hdf1":  # Why special case?

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -71,19 +71,21 @@ def custom_ad_creator(plugin_name):
     # fmt: on
 )
 def test_AD_EpicsFileNameMixin(plugin_name, spec):
+    import epics
+
     adsimdet = custom_ad_creator(plugin_name)
     assert adsimdet is not None
     assert plugin_name in dir(adsimdet)
 
     # basic plugin tests
-    mixin = getattr(adsimdet, plugin_name)
+    plugin = getattr(adsimdet, plugin_name)
     assert AD_EpicsFileNameMixin is not None
-    assert isinstance(mixin, AD_EpicsFileNameMixin)
-    assert "filestore_spec" in dir(mixin)
-    assert mixin.filestore_spec == spec
-    assert isinstance(mixin.stage_sigs, dict)
-    assert len(mixin.stage_sigs) >= 4
-    assert "capture" in mixin.stage_sigs
+    assert isinstance(plugin, AD_EpicsFileNameMixin)
+    assert "filestore_spec" in dir(plugin)
+    assert plugin.filestore_spec == spec
+    assert isinstance(plugin.stage_sigs, dict)
+    assert len(plugin.stage_sigs) >= 4
+    assert "capture" in plugin.stage_sigs
 
     # configuration prescribed for the user
     user_settings = dict(
@@ -101,20 +103,21 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec):
         user_settings["compression"] = "zlib"
 
     # add the settings
-    for k, v in user_settings.items():
-        getattr(mixin, k).put(v)
+    for attr, value in user_settings.items():
+        sig = getattr(plugin, attr)
+        epics.caput(sig._write_pv.pvname, value)
     timed_pause()
 
     adsimdet.stage()
     if plugin_name == "hdf1":  # Why special case?
         user_settings["file_number"] += 1  # the IOC will do the same
-    assert list(mixin.stage_sigs.keys())[-1] == "capture"
+    assert list(plugin.stage_sigs.keys())[-1] == "capture"
     for k, v in user_settings.items():
-        assert getattr(mixin, k).get() == v, f"{plugin_name=} {k=}  {v=}"
+        assert getattr(plugin, k).get() == v, f"{plugin_name=} {k=}  {v=}"
 
-    assert mixin.get_frames_per_point() == user_settings["num_capture"]
+    assert plugin.get_frames_per_point() == user_settings["num_capture"]
 
-    filename, read_path, write_path = mixin.make_filename()
+    filename, read_path, write_path = plugin.make_filename()
     assert isinstance(filename, str)
     assert isinstance(read_path, str)
     assert isinstance(write_path, str)

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -93,10 +93,11 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec):
         auto_increment=1,  # Yes
         auto_save=0,  # No
         create_directory=-5,
+        # file_template before file_name & file_path
+        file_template=f"%s%s_%2.2d.{plugin_name}",
         file_name="flotsam",
         file_number=1 + int(10 * random.random()),
         file_path=f"{pathlib.Path(tempfile.mkdtemp())}/",  # ALWAYS ends with "/"
-        file_template=f"%s%s_%2.2d.{plugin_name}",
         num_capture=1 + int(10 * random.random()),
     )
     if plugin_name == "hdf1":

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -74,10 +74,12 @@ def custom_ad_creator(plugin_name):
     [
         [
             dict(
+                auto_save="No",
+                lazy_open="No",
                 create_directory=-5,
                 file_template="%s%s_%2.2d.ad_data",
-                file_name="test",
                 file_path="/tmp/",
+                file_name="test",
             ),
             does_not_raise(),
             None,
@@ -100,7 +102,14 @@ def test_replacement(plugin_name, spec, presets, context, expected):
         assert plugin.filestore_spec == spec
         assert isinstance(plugin.stage_sigs, dict)
         assert len(plugin.stage_sigs) >= 4
-        assert "capture" in plugin.stage_sigs
+        assert plugin.stage_sigs.get("file_write_mode") == "Stream"
+        assert plugin.stage_sigs.get("capture") in (1, "Capture")
+        plugin.stage_sigs.move_to_end("capture", last=True)
+        assert list(plugin.stage_sigs)[-1] == "capture"
+
+        if not AD_plugin_primed(plugin):
+            AD_prime_plugin2(plugin)
+            assert AD_plugin_primed(plugin)
 
         settings = presets.copy()  # Do NOT change the 'presets' parameter.
         if plugin_name == "hdf1":

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -13,7 +13,10 @@ import bluesky
 import pytest
 from bluesky import plan_stubs as bps
 from bluesky import plans as bp
+from ophyd.utils.errors import UnknownStatusFailure
+from ophyd.utils.errors import UnprimedPlugin
 
+from ...tests import IOC_AD
 from ...tests import READ_PATH_TEMPLATE
 from ...tests import MonitorCache
 from ...tests import timed_pause
@@ -22,7 +25,10 @@ from .. import AD_EpicsFileNameJPEGPlugin
 from .. import AD_EpicsFileNameMixin
 from .. import AD_EpicsFileNameTIFFPlugin
 from .. import AD_full_file_name_local
+from .. import AD_plugin_primed
+from .. import AD_prime_plugin2
 from .. import BadPixelPlugin
+from .. import ad_creator
 
 
 @pytest.mark.parametrize(
@@ -309,6 +315,142 @@ def test_plugin(Klass, setup, attrs, context, expected):
         knowns = dir(obj)
         for attr in attrs:
             assert attr in knowns  # hasattr() needs connection
+
+    if expected is not None:
+        assert expected in str(exinfo)
+
+
+@pytest.mark.parametrize(
+    "setup",
+    [
+        dict(
+            plugins=[
+                "cam",
+                dict(
+                    hdf1={
+                        "class": AD_EpicsFileNameHDF5Plugin,
+                        "read_path_template": "/tmp/docker_ioc/iocad/tmp/",
+                        "write_path_template": "/tmp/",
+                    }
+                ),
+            ]
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "attrs, context, expected",
+    [
+        [
+            dict(
+                auto_save="No",
+                create_directory=0,
+                file_name="",
+                file_number=1,
+                file_path="",
+                file_template="",
+            ),
+            pytest.raises(ValueError),
+            "'ad:HDF1:FileName' is an empty string.",
+        ],
+        [
+            dict(
+                auto_save="No",
+                create_directory=-5,
+                file_template="",
+                file_name="test",
+                file_number=1,
+                file_path="",
+            ),
+            pytest.raises(ValueError),
+            "'ad:HDF1:FilePath' is an empty string.",
+        ],
+        [
+            dict(
+                auto_save="No",
+                create_directory=-5,
+                file_template="",
+                file_name="test",
+                file_number=1,
+                file_path="/tmp/",
+            ),
+            pytest.raises(ValueError),
+            "'ad:HDF1:FileTemplate' is an empty string.",
+        ],
+        [
+            dict(
+                auto_save="No",
+                create_directory=-5,
+                file_template="%s%s_%6.6d.h5",
+                file_name="test",
+                file_number=1,
+                file_path="/tmp/",
+            ),
+            does_not_raise(),
+            None,
+        ],
+        [
+            dict(
+                auto_save="No",
+                create_directory=-5,
+                file_template="%s%s_%6.6d.h5",
+                file_name="test",
+                file_number=1,
+                file_path="/",
+            ),
+            pytest.raises(OSError),
+            "Path '/' does not exist on IOC",
+        ],
+    ],
+)
+def test_HDF5plugin_i1062(setup, attrs, context, expected):
+    """Issue #1062, Unconfigured HDF plugin."""
+    import epics
+
+    # Caller is responsible for setting values of these Components:
+    # * array_counter
+    # * auto_increment
+    # * auto_save
+    # * compression (only HDF)
+    # * create_directory
+    # * file_name
+    # * file_number
+    # * file_path
+    # * file_template
+    # * num_capture
+
+    with context as exinfo:
+        det = ad_creator(IOC_AD, name="det", **setup)
+
+        # Confirm these attributes are not staged.
+        for attr in "file_name file_path file_template".split():
+            assert attr not in det.hdf1.stage_sigs
+
+        # # un-prime the HDF5 plugin
+        # det.cam.data_type.put("Int8")
+        # AD_prime_plugin2(det.hdf1)
+        # det.cam.data_type.put("UInt8")
+        # time.sleep(0.1)
+        # assert not AD_plugin_primed(det.hdf1)
+
+        if not AD_plugin_primed(det.hdf1):
+            AD_prime_plugin2(det.hdf1)
+            assert AD_plugin_primed(det.hdf1)
+
+        for attr, value in attrs.items():
+            epics.caput(getattr(det.hdf1, attr)._write_pv.pvname, value)
+
+        # # This happens during det.hdf1.stage()
+        # filename, read_path, write_path = det.hdf1.make_filename()
+        # # Test its results.
+        # formatter = datetime.datetime.now().strftime
+        # assert filename == det.hdf1.file_name.get()
+        # assert read_path == setup["plugins"][1]["hdf1"]["read_path_template"]
+        # assert read_path == formatter(det.hdf1.read_path_template)
+        # assert write_path == formatter(det.hdf1.file_path.get())
+
+        RE = bluesky.RunEngine()
+        (uid,) = RE(bp.count([det]))
+        assert isinstance(uid, str)
 
     if expected is not None:
         assert expected in str(exinfo)

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -18,6 +18,7 @@ from ...tests import IOC_AD
 from ...tests import READ_PATH_TEMPLATE
 from ...tests import WRITE_PATH_TEMPLATE
 from ...tests import MonitorCache
+from ...tests import in_gha_workflow
 from ...tests import timed_pause
 from .. import AD_EpicsFileNameHDF5Plugin
 from .. import AD_EpicsFileNameJPEGPlugin
@@ -61,21 +62,91 @@ def custom_ad_creator(plugin_name):
 
 
 @pytest.mark.parametrize(
-    # fmt: off
     "plugin_name, spec",
     [
         ["hdf1", "AD_HDF5"],
         ["jpeg1", "AD_JPEG"],
         ["tiff1", "AD_TIFF"],
     ],
-    # fmt: on
 )
-def test_AD_EpicsFileNameMixin(plugin_name, spec):
+@pytest.mark.parametrize(
+    "presets, context, expected",
+    [
+        [
+            dict(
+                create_directory=-5,
+                file_template="%s%s_%2.2d.ad_data",
+                file_name="test",
+                file_path="/tmp/",
+            ),
+            does_not_raise(),
+            None,
+        ],
+    ],
+)
+def test_replacement(plugin_name, spec, presets, context, expected):
     import epics
 
-    adsimdet = custom_ad_creator(plugin_name)
+    def caput(signal, value):
+        epics.caput(signal._write_pv.pvname, value)
+
+    with context as exinfo:
+        det = custom_ad_creator(plugin_name)
+
+        plugin = getattr(det, plugin_name)
+        assert AD_EpicsFileNameMixin is not None
+        assert isinstance(plugin, AD_EpicsFileNameMixin)
+        assert "filestore_spec" in dir(plugin)
+        assert plugin.filestore_spec == spec
+        assert isinstance(plugin.stage_sigs, dict)
+        assert len(plugin.stage_sigs) >= 4
+        assert "capture" in plugin.stage_sigs
+
+        settings = presets.copy()  # Do NOT change the 'presets' parameter.
+        if plugin_name == "hdf1":
+            settings["compression"] = "zlib"
+
+        for attr, value in settings.items():
+            if attr in dir(plugin):
+                caput(getattr(plugin, attr), value)
+
+        RE = bluesky.RunEngine()
+        (uid,) = RE(bp.count([det]))
+        assert isinstance(uid, str)
+
+    if expected is not None:
+        assert expected in str(exinfo)
+
+
+@pytest.mark.skipif(
+    in_gha_workflow(),
+    reason="Random failures in GiHub Actions workflows.",
+)
+@pytest.mark.parametrize(
+    "plugin_name, spec",
+    [
+        ["hdf1", "AD_HDF5"],
+        ["jpeg1", "AD_JPEG"],
+        ["tiff1", "AD_TIFF"],
+    ],
+)
+def test_AD_EpicsFileNameMixin(plugin_name, spec, adsimdet):
+    import epics
+
+    def caput(signal, value):
+        epics.caput(signal._write_pv.pvname, value)
+
+    # adsimdet = custom_ad_creator(plugin_name)
     assert adsimdet is not None
     assert plugin_name in dir(adsimdet)
+
+    # Preset some plugin values required for staging.
+    for key in "hdf jpeg tiff".split():
+        plugin = getattr(adsimdet, f"{key}1")
+        caput(plugin.create_directory, -5)
+        caput(plugin.file_name, f"test_{key}")
+        caput(plugin.file_path, f"/tmp/test_{key}/")
+        caput(plugin.file_template, f"%s%s_%3.3d.{key}/")
 
     # basic plugin tests
     plugin = getattr(adsimdet, plugin_name)
@@ -105,8 +176,7 @@ def test_AD_EpicsFileNameMixin(plugin_name, spec):
 
     # add the settings
     for attr, value in user_settings.items():
-        sig = getattr(plugin, attr)
-        epics.caput(sig._write_pv.pvname, value)
+        caput(getattr(plugin, attr), value)
     timed_pause(0.1)
 
     adsimdet.stage()

--- a/apstools/devices/tests/test_area_detector_support.py
+++ b/apstools/devices/tests/test_area_detector_support.py
@@ -13,11 +13,10 @@ import bluesky
 import pytest
 from bluesky import plan_stubs as bps
 from bluesky import plans as bp
-from ophyd.utils.errors import UnknownStatusFailure
-from ophyd.utils.errors import UnprimedPlugin
 
 from ...tests import IOC_AD
 from ...tests import READ_PATH_TEMPLATE
+from ...tests import WRITE_PATH_TEMPLATE
 from ...tests import MonitorCache
 from ...tests import timed_pause
 from .. import AD_EpicsFileNameHDF5Plugin
@@ -31,6 +30,36 @@ from .. import BadPixelPlugin
 from .. import ad_creator
 
 
+def custom_ad_creator(plugin_name):
+    """Create area detector with named file writer plugin."""
+    plugin_class = dict(
+        hdf1=AD_EpicsFileNameHDF5Plugin,
+        jpeg1=AD_EpicsFileNameJPEGPlugin,
+        tiff1=AD_EpicsFileNameTIFFPlugin,
+    )
+    writer_plugin = {
+        plugin_name: {
+            "class": plugin_class[plugin_name],
+            "write_path_template": WRITE_PATH_TEMPLATE,
+            "read_path_template": READ_PATH_TEMPLATE,
+        }
+    }
+    if plugin_name in "jpeg1 tiff1".split():
+        writer_plugin[plugin_name]["suffix"] = f"{plugin_name.upper()}:"
+    adsimdet = ad_creator(
+        IOC_AD,
+        name="adsimdet",
+        class_name="MySimDet",
+        plugins=[
+            "cam",
+            "image",
+            "pva",
+            writer_plugin,
+        ],
+    )
+    return adsimdet
+
+
 @pytest.mark.parametrize(
     # fmt: off
     "plugin_name, spec",
@@ -41,7 +70,8 @@ from .. import ad_creator
     ],
     # fmt: on
 )
-def test_AD_EpicsFileNameMixin(plugin_name, spec, adsimdet):
+def test_AD_EpicsFileNameMixin(plugin_name, spec):
+    adsimdet = custom_ad_creator(plugin_name)
     assert adsimdet is not None
     assert plugin_name in dir(adsimdet)
 
@@ -127,8 +157,10 @@ def test_stage(plugin_name, plugin_class, adsimdet, fname):
         ["tiff1", AD_EpicsFileNameTIFFPlugin],
     ],
 )
-def test_bp_count_custom_name(plugin_name, plugin_class, adsimdet, fname):
+def test_bp_count_custom_name(plugin_name, plugin_class, fname):
     assert isinstance(plugin_name, str)
+
+    adsimdet = custom_ad_creator(plugin_name)
 
     plugin = getattr(adsimdet, plugin_name)
     assert isinstance(plugin, plugin_class)
@@ -148,7 +180,7 @@ def test_bp_count_custom_name(plugin_name, plugin_class, adsimdet, fname):
             adsimdet.cam.acquire_time, 0.001,
             adsimdet.cam.acquire_period, 0.002,
             adsimdet.cam.image_mode, "Single",
-            # plugin.file_template  # pre-configured
+            plugin.file_template, "%s%s_%3.3d.h5"
         )
         # fmt: on
 
@@ -195,7 +227,8 @@ def test_bp_count_custom_name(plugin_name, plugin_class, adsimdet, fname):
     assert fname in str(lfname), f"{lfname=}  {fname=}"
 
 
-def test_full_file_name_local(adsimdet, fname):
+def test_full_file_name_local(fname):
+    adsimdet = custom_ad_creator("hdf1")
     plugin = adsimdet.hdf1
     plugin.file_name.put(fname)
     plugin.file_path.put(f"{plugin.write_path_template}/")
@@ -211,23 +244,8 @@ def test_full_file_name_local(adsimdet, fname):
     assert str(lfname).find(fname) > 0
 
 
-def test_no_file_path(adsimdet, fname):
-    adsimdet.hdf1.file_name.put(fname)
-    adsimdet.hdf1.file_path.put("/no-such-path-exists")
-    time.sleep(0.1)
-    assert adsimdet.hdf1.file_path_exists.get() == 0
-
-    adsimdet.hdf1.file_path.put("")
-    time.sleep(0.1)
-    assert adsimdet.hdf1.file_path_exists.get() == 0
-
-    with pytest.raises(OSError) as exc:
-        adsimdet.stage()
-    assert str(exc.value).endswith(" does not exist on IOC.")
-    adsimdet.unstage()
-
-
-def test_file_numbering(adsimdet, fname):
+def test_file_numbering(fname):
+    adsimdet = custom_ad_creator("hdf1")
     plugin = adsimdet.hdf1
     plugin.file_name.put(fname)
     plugin.file_path.put(f"{plugin.write_path_template}/")
@@ -262,7 +280,9 @@ def test_file_numbering(adsimdet, fname):
 
 
 def test_capture_error_with_single_mode(adsimdet):
+    # adsimdet = custom_ad_creator("hdf1")
     plugin = adsimdet.hdf1
+    plugin.file_name.put("test")
 
     # test that plugin is in the wrong configuration for Single mode
     assert "capture" in plugin.stage_sigs

